### ISSUE TITLE
focus-visible doesn't work in Firefox even with polyfill

### ISSUE THREAD

**Issue by jankalfus** [tailwindlabs/tailwindcss#3325]
### Describe the problem:
Using the classes prefixed by `focus-visible:`, like `focus-visible:ring-2` doesn't work at all. No focus ring is present on the element when focused by keyboard. It isn't working even in the docs at https://tailwindcss.com/docs/hover-focus-and-other-states#focus-visible

![focus-visible-tailwind](https://user-images.githubusercontent.com/12714276/104134222-91b2bd80-5388-11eb-9a67-33f7271bfe51.gif)

Note this isn't issue with the docs, this is an issue with tailwind. I've also set up tailwind + focus-visible js polyfill and postcss plugin locally and I'm observing the same behavior - no focus ring at all. I've confirmed that the polyfill is working in the docs (and also locally) by looking at the elements structure.

Browser: Firefox 84.0.2 (note that v85 is probably shipping soon, which should have native focus-visible support)

### Link to a minimal reproduction:

Go to https://tailwindcss.com/docs/hover-focus-and-other-states#focus-visible in Firefox 84, try tabbing and observe that there's no focus ring for the "Ring on focus-visible" button.

**Comment by adamwathan** [comment]
Hey! It actually does work, here's a sample project:

https://github.com/adamwathan/focus-visible-test

And a preview URL:

https://affectionate-wing-d0d0f9.netlify.app/

And a video:

https://user-images.githubusercontent.com/4323180/104135643-68316c00-535f-11eb-90a2-cfdfe7dabac5.mov

We had a bug on the actual Tailwind docs (we recently rebuilt it) where we forgot to add the `postcss-focus-visible` polyfill but I've just added that and it's working as expected there as well. Thanks for catching that.

Closing this as I've demonstrated it works with the example project above but happy to help troubleshoot the configuration issue in your project. The example repo above should be enough to figure out what the configuration difference is though 👍🏻

**Comment by jankalfus** [comment]
I checked the sample project. Seems like I had to add the following to my `tailwind.config.js` to make it work:

```
  variants: {
    extend: {
      ringWidth: ['focus-visible'],
      ringColor: ['focus-visible'],
      ringOffsetWidth: ['focus-visible'],
      ringOffsetColor: ['focus-visible'],
    },
```

Thanks!

Shouldn't this be mentioned somewhere in the docs though?

**Comment by adamwathan** [comment]
It is mentioned in a few spots, one in the focus-visible docs:

<img width="877" alt="Screen Shot 2021-01-11 at 1 26 03 PM" src="https://user-images.githubusercontent.com/4323180/104222698-a89cf280-5410-11eb-83ba-c8914f4e3cf0.png">

And also in the "Variants" section at the bottom of every single utility page, including [Ring Width](https://tailwindcss.com/docs/ring-width#variants) and the other ring utilities:

![image](https://user-images.githubusercontent.com/4323180/104222778-c36f6700-5410-11eb-927d-e79ac138917d.png)

Is there anywhere you would expect to see it that we're missing it?

**Comment by jankalfus** [comment]
Hmm, I saw that text in bold with the example. But I guess the issue is that as a newcommer to tailwind, I don't really know what the "By default... not enabled for any core plugins" sentence really means. I've read pieces of documentation here and there, but I have no idea that tailwind uses plugins or what those are. Pardon my ignorance, but I don't believe it's something a newcommer who just wants to use focus-visible in combination with some basic styling should worry about. In comparison, setting up the focus-visible polyfill without tailwind is pretty straightforward.

I also don't know that some variants can be generated and some not, heck I don't even know what a "variant" is 😄, I don't realize seeing it in the first free pages of the core concepts docs section. I'd expect that all are available by default and those I'm not using would get detected and removed by the post processor. Is that too naive? 😄 

I've been using the docs as a reference, rather than a book that I'd read cover to cover. I'm guessing a link to some explanation what the plugins are might be helpful. And if the plugins are really such essential concept they seem to be, I'm guessing it'd be really nice to mention it right in the installation where the "Learn more about configuring Tailwind in the configuration documentation." sentence is on this page: https://tailwindcss.com/docs/guides/vue-3-vite Maybe something like "Not all CSS classes are available by default, check the configuration docs to see how to enable them." would attract my attention.

In general, I guess tailwind looks super easy to use for somebody new coming to the framework, and the more surprising it is to see it actually isn't. That's probably the issue I'm having 😢

**Comment by lolcode8** [comment]
@adamwathan Do you know why the script and the npm package are required separately here? Works great on Firefox and Chrome, trying to get it to work on Safari. Curious why the extra steps.

**Comment by rapritchard** [comment]
I know this thread is closed, but I've followed the docs exactly and focus-visible is still refusing to work.

I've added the polyfills and set up the tailwind config to enable the variants.

**Comment by beepy** [comment]
Replying to the question, "is there anywhere you'd expect to see [the mention of enabling `focus-visible` in the core plugins] that we’re missing it?"

I think the confusion is that the example given below the "by default, the focus-visible variant is not enabled for any core plugins" statement doesn't cover a typical use case; jankalfus's discovered sample is more typically the behavior a person would be trying to enable, that is, causing a focus ring to appear for the `:focus-visible` pseudo class.

**Comment by AlexMartinFR** [comment]
For those who have doubts about which variants you need to configure, I'd recommend trying out [Tailwind's JIT mode](https://tailwindcss.com/docs/just-in-time-mode).

Besides speed and file-size advantages, you also don't need to configure variants anymore.
This means you can use `focus-visible:` straight out of the box 👍

**Comment by sidgoel85** [comment]
> @adamwathan Do you know why the script and the npm package are required separately here? Works great on Firefox and Chrome, trying to get it to work on Safari. Curious why the extra steps.

Hi There, Just curious if you we're able to make that work in Safari, We're having exact same issue where focus-visible working in Chrome/Firefox but refuses to work in Safari.
